crates/egui/src/text_selection/text_cursor_state.rs

```diff
@@ -302,18 +302,108 @@ pub fn find_line_start(text: &str, current_index: CCursor) -> CCursor {
     // number of multi byte chars.
     // We need to know the char index to be able to correctly set the cursor
     // later.
-    let chars_count = text.chars().count();
+    let mut char_line_indexes: Vec<usize> = create_char_line_indexes(text);
+    let line_start_index = get_line_start_index(&mut char_line_indexes, current_index.index);
 
-    let position = text
-        .chars()
-        .rev()
-        .skip(chars_count - current_index.index)
-        .position(|x| x == '\n');
+    CCursor::new(line_start_index)
+}
+
+pub fn create_char_line_indexes(text: &str) -> Vec<usize> {
+    let mut line_indexes = vec![0];
+
+    let mut count = 0;
+    for c in text.chars() {
+        count += 1;
+        if c == '\n' {
+            line_indexes.push(count);
+        }
+    }
+    line_indexes.push(count);
+
+    line_indexes
+}
+
+pub fn get_line_start_index(line_indexes: &mut [usize], index: usize) -> usize {
```

Comment:
There is a lot of naked `usize`. If you use `CCursor` you can make it clear wether or not something is a byte offset or a character offset

Reply:
With this, we can easily find out the current Colume and line within TextEdit. I'm using this, and I'm sure others might want to use it too.

Reply:
does it return a byte index or a char index? The name doesn't tell me, neither does the return type